### ISSUE TITLE
[Flaky Test]: TestStandaloneUpgrade/Upgrade_8.12.0-SNAPSHOT_to_8.13.0-SNAPSHOT_(privileged) – error installing package: failed to copy source directory... /opt/Elastic/Agent/elastic-agent: file exists

### ISSUE THREAD

**Issue by pchila** [elastic/elastic-agent#4204]
### Failing test case

TestStandaloneUpgrade/Upgrade_8.12.0-SNAPSHOT_to_8.13.0-SNAPSHOT_(privileged)

### Error message

/opt/Elastic/Agent/elastic-agent: file exists during install, looks like a previous install was not cleaned or is still being cleaned

### Build

https://buildkite.com/elastic/elastic-agent/builds/6855#018d7f0a-a775-42a3-aa0b-0fb496da1a97

### OS

Linux

### Stacktrace and notes

```markdown
fixture.go:631: >> running binary with: [/tmp/TestStandaloneUpgradeUpgrade_8.12.0-SNAPSHOT_to_8.13.0-SNAPSHOT_(privileged)332107157/001/elastic-agent-8.12.0-SNAPSHOT-linux-arm64/elastic-agent install --force --non-interactive]
    upgrade_standalone_test.go:77: 
        	Error Trace:	/home/ubuntu/agent/testing/integration/upgrade_standalone_test.go:77
        	            				/home/ubuntu/agent/testing/integration/upgrade_standalone_test.go:52
        	Error:      	Received unexpected error:
        	            	failed to install start agent (err: error running agent install command: exit status 1) [output: Installing in non-interactive mode.
        	            	
[    ] Copying install files  [0s] 
                                   

[   =] Error copying files  [12s] Error: error installing package: failed to copy source directory (/tmp/TestStandaloneUpgradeUpgrade_8.12.0-SNAPSHOT_to_8.13.0-SNAPSHOT_(privileged)332107157/001/elastic-agent-8.12.0-SNAPSHOT-linux-arm64) to destination (/opt/Elastic/Agent): symlink data/elastic-agent-bf5180/elastic-agent /opt/Elastic/Agent/elastic-agent: file exists
        	            	For help, please see our troubleshooting guide at https://www.elastic.co/guide/en/fleet/8.12/fleet-troubleshooting.html
        	            	]
        	Test:       	TestStandaloneUpgrade/Upgrade_8.12.0-SNAPSHOT_to_8.13.0-SNAPSHOT_(privileged)
--- FAIL: TestStandaloneUpgrade/Upgrade_8.12.0-SNAPSHOT_to_8.13.0-SNAPSHOT_(privileged) (56.84s)
```

There are more tests failing in the same build with the same root cause:
- TestStandaloneUpgradeUninstallKillWatcher
- TestStandaloneUpgradeRollback
- TestStandaloneUpgrade/Upgrade_8.12.1-SNAPSHOT_to_8.13.0-SNAPSHOT_(privileged)

Also in the same build on the same runner, test `TestStandaloneUpgradeRetryDownload` fails because the agent disappears mid-test:

```
  upgrader.go:363: waiting for upgrade watcher to finish
    upgrader.go:368: upgrade watcher finished
    upgrader.go:373: Checking upgrade details state after successful upgrade
    fixture.go:631: >> running binary with: [/opt/Elastic/Agent/elastic-agent version --binary-only --yaml]
    upgrade_standalone_retry_test.go:111: 
        	Error Trace:	/home/ubuntu/agent/testing/integration/upgrade_standalone_retry_test.go:111
        	Error:      	Received unexpected error:
        	            	failed to get Agent version: fork/exec /opt/Elastic/Agent/elastic-agent: no such file or directory
        	Test:       	TestStandaloneUpgradeRetryDownload
    fixture_install.go:165: [test TestStandaloneUpgradeRetryDownload] Inside fixture cleanup function
    fixture_install.go:180: collecting diagnostics; test failed
    fixture.go:631: >> running binary with: [/opt/Elastic/Agent/elastic-agent diagnostics -f /home/ubuntu/agent/build/diagnostics/TestStandaloneUpgradeRetryDownload-diagnostics-2024-02-06T16:11:30Z.zip]
    fixture_install.go:306: failed to collect diagnostics to /home/ubuntu/agent/build/diagnostics/TestStandaloneUpgradeRetryDownload-diagnostics-2024-02-06T16:11:30Z.zip (fork/exec /opt/Elastic/Agent/elastic-agent: no such file or directory): 
    fixture_install.go:319: creating zip archive of the installation directory: /opt/Elastic/Agent
    fixture_install.go:349: failed to add /opt/Elastic/Agent/elastic-agent to zip, continuing: open /opt/Elastic/Agent/elastic-agent: no such file or directory
    fixture_install.go:349: failed to add /opt/Elastic/Agent/elastic-agent.sock to zip, continuing: open /opt/Elastic/Agent/elastic-agent.sock: no such device or address
    fixture.go:631: >> running binary with: [/opt/Elastic/Agent/elastic-agent uninstall --force]
    fixture_install.go:209: fixture.Install Cleanup: agent was already uninstalled, skipping uninstall
```

This seems like a slow watcher or another slow test cleaning up...

It's curious as it only seem to happen on the arm64 runner
```

**Comment by cmacknz** [comment]
@pchila have you seen this happen on main? Or is this something isolated to your feature branch since that is where the report is from?

From memory I don't recognize this error as something that has happened before.

**Comment by pchila** [comment]
I am seeing this pretty consistently on builds for PR https://github.com/elastic/elastic-agent/pull/4193
It seems to happen only on the ubuntu arm64 runner and started happening after a rebase on main yesterday.
My suspicion is that this is not strictly related to the PR changes (I would expect to see failures in more runners, at least also in the linux amd64 one) but my PR is bringing this out consistently.

It has all the markings of a rogue watcher that keeps running, does not get removed, block further agent installs as the symlink is still present and then it will asynchronously remove everything. From that moment on tests start passing again

**Comment by cmacknz** [comment]
> It has all the markings of a rogue watcher that keeps running

We should probably starting failing tests if we find an `elastic-agent watch` process still running before a new test starts or after we execute `elastic-agent uninstall`. The uninstall command is supposed to be removing it, so we need to check this after that command executes.

**Comment by pchila** [comment]
Used https://github.com/elastic/elastic-agent/pull/4217 to run CI adding extra assertions about elastic-agent processes still running before/after a test. It points to `TestStandaloneDowngradeToSpecificSnapshotBuild` leaving a process that keeps other tests from installing at first and then asynchronously cleans up agent installations

**Comment by rdner** [comment]
@pchila One note: I'd prefer if we skipped tests as separate PRs, not as a part of a bigger change where it can get lost. Also, unskipping would be as easy as reverting a PR.

Regarding `TestStandaloneDowngradeToSpecificSnapshotBuild`, it might be the reason why we have uninstall failures on Windows as well.

I don't see anything special about this particular tests and how it runs the upgrade. Other tests, for example [this test case](https://github.com/elastic/elastic-agent/blob/5ce913edf4835dd3b90f1de3a354568919c3145b/testing/integration/upgrade_standalone_retry_test.go#L109-L110) runs it very similarly and does not have any additional cleanup.

Are we sure this test is consistently leaving the process running? Or it's rather our testing framework that can do that for any test case randomly?

**Comment by pchila** [comment]
Fixed on feature branch by #4229. Merged on main through #4193
